tests/unit/s2n_ktls_test.c

```diff
@@ -38,5 +78,179 @@ int main(int argc, char **argv)
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_recv(server_conn), S2N_ERR_HANDSHAKE_NOT_COMPLETE);
+        };
+
+        /* s2n_connection_ktls_enable */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            int fd = 1;
+            EXPECT_OK(s2n_test_configure_mock_ktls_connection(server_conn, fd, true));
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_send(server_conn), S2N_ERR_KTLS_DISABLED_FOR_TEST);
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_recv(server_conn), S2N_ERR_KTLS_DISABLED_FOR_TEST);
+        };
+
+        /* ktls already enabled is a noop and returns success */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            int fd = 1;
+            EXPECT_OK(s2n_test_configure_mock_ktls_connection(server_conn, fd, true));
+
+            server_conn->ktls_send_enabled = true;
+            EXPECT_SUCCESS(s2n_connection_ktls_enable_send(server_conn));
+
+            server_conn->ktls_recv_enabled = true;
+            EXPECT_SUCCESS(s2n_connection_ktls_enable_recv(server_conn));
+        };
+
+        /* unsupported protocols */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            int fd = 1;
+            EXPECT_OK(s2n_test_configure_mock_ktls_connection(server_conn, fd, true));
+
+            server_conn->actual_protocol_version = S2N_TLS13;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_send(server_conn), S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+            server_conn->actual_protocol_version = S2N_TLS11;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_send(server_conn), S2N_ERR_KTLS_UNSUPPORTED_CONN);
+        };
+
+        /* unsupported ciphers */
+        {
+            /* set kTLS un-supported cipher */
+            struct s2n_cipher ktls_temp_unsupported_cipher = {
+                .ktls_supported = false,
+            };
+            struct s2n_record_algorithm ktls_temp_unsupported_record_alg = {
+                .cipher = &ktls_temp_unsupported_cipher,
+            };
+            struct s2n_cipher_suite ktls_temp_unsupported_cipher_suite = {
+                .record_alg = &ktls_temp_unsupported_record_alg,
+            };
+
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            int fd = 1;
+            EXPECT_OK(s2n_test_configure_mock_ktls_connection(server_conn, fd, true));
+
+            /* base case */
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_ktls_enable_send(server_conn), S2N_ERR_KTLS_DISABLED_FOR_TEST);
```

Comment:
I'm torn about this. I don't think this error should be your "successful" case. It's confusing to look through this test and see how the comments contradict the code. What if instead the ktls code nooped instead of erroring? Then you could actually expect this function to succeed.